app/javascript/serviceworker/store.ts

```diff
@@ -1,58 +1,78 @@
 import { openDB, DBSchema } from "idb";
 
+type StoreName = "delayedRequests";
+
 interface RequestObject {
   id?: number;
   url: string;
-  data: any;
+  body: any;
 }
 
 interface OfflineDatabase extends DBSchema {
-  requests: {
+  delayedRequests: {
     key: number;
     value: RequestObject;
+    indexes: { url: string };
+  };
+  cachedResponses: {
+    key: number;
+    value: RequestObject;
+    indexes: { url: string };
   };
-}
-
-interface Db {
-  createObjectStore: (
-    arg0: string,
-    arg1: { keyPath: string; autoIncrement: boolean }
-  ) => void;
 }
 
 const DB_NAME = "offline";
 const DB_VERSION = 1;
 
-const openTx = async (mode: "readwrite" | "readonly") => {
+const openTx = async (storeName: StoreName, mode: "readwrite" | "readonly") => {
   const db = await openDB<OfflineDatabase>(DB_NAME, DB_VERSION, {
-    upgrade(db: Db) {
-      db.createObjectStore("requests", {
+    upgrade(db) {
+      const delayedStore = db.createObjectStore("delayedRequests", {
         keyPath: "id",
         autoIncrement: true,
       });
+
+      delayedStore.createIndex("url", "url", { unique: false });
+
+      const cachedStore = db.createObjectStore("cachedResponses", {
+        keyPath: "id",
+        autoIncrement: true,
+      });
+
+      cachedStore.createIndex("url", "url", { unique: true });
     },
   });
 
-  return db.transaction("requests", mode);
+  return db.transaction(storeName, mode);
 };
 
-export const saveRequest = async (request: Request) => {
-  const requestData = await request.formData();
-  const data = Object.fromEntries(requestData);
-
-  const tx = await openTx("readwrite");
-  await tx.store.add({ url: request.url, data });
+export const add = async (storeName: StoreName, url: string, body: any) => {
```

Comment:
It feels odd to me here, and for the funcs below, that the caller would need to know the name of the db being used ... can we have a chat to hash this out?

Reply:
Works. Nice!